### DISCUSSION TITLE
quarkus api secured with Azure AD

### DISCUSSION THREAD

**Discussion by dkoci** [quarkusio/quarkus#25192] [Community]
Hi folks.
Does anyone have some experience with quarkus api which use azure ad for authentication?
I have backend (quarkus api) and frontend (angular) which calls quarkus api. I have found some springboot examples, but it is not working with quarkus-oidc extension. I have googled alot, but more or less there are only quarkus/keycloak exaples.

Thank you

**Comment by sberyozkin** [comment]
Hi, I believe we've had many users confirming it works with Azure AD. There is no much difference between using Keycloak and Azure with `quarkus-oidc`. I'm assuming it is your frontend application which is actually authenticating the users, in which case all you need is to include the access token returned from Azure to invoke on Quarkus, use `Authorization: Bearer <token>` - I think just adding `quarkus.oidc.auth-server-url=azure-address` should be enough in this case, where `azure-address` identifies the Azure endpoint from where the verification keys and other OIDC info can be fetched from.

**Reply by dkoci** [reply]
Thank you for your answer. The problem was in server url. I had `quarkus.oidc.auth-server-url=https://login.microsoftonline.com/<tenant-id>/v2.0` which works with **idtoken**. But my frontend sends **accesstoken** to backend. The result was 401: unauthorized.
It must be `quarkus.oidc.auth-server-url=https://sts.windows.net/<tenant-id>` which accepts access-token.

**Comment by sberyozkin** [comment]
@dkoci Right, this is because the access token provided by Azure is binary (not JWT), if it were in JWT format then it would work as well. 
`https://login.microsoftonline.com/<tenant-id>/v2.0` would work if Quarkus itself were doing the code flow.
I did not know that `quarkus.oidc.auth-server-url=https://sts.windows.net/<tenant-id>` is where one can verify the opaque binary Azure tokens, it is a useful info which we can share with other Quarkus Azure users, thanks

**Reply by dkoci** [reply]
I found it here: https://ordina-jworks.github.io/security/2020/08/18/Securing-Applications-Azure-AD.html
It is springboot api but the same settings (application.properties) works with quarkus too. It is also explained.

**Reply by mariuszbozek** [reply]
Hi, 
I use OIDC extension with Azure AD
My steps:
1. register my backend service in Azure AD (App Registration)
2. create clientId and secret. 
3. expose App Registration (expose API)
4. add properties
```
quarkus.oidc.client-id=
quarkus.oidc.credentials.secret=
quarkus.oidc.auth-server-url=https://login.microsoftonline.com/<tenant>/v2.0
quarkus.oidc.application-type=service
quarkus.oidc.authentication.redirect-path=/admin
quarkus.oidc.authentication.scopes=profile
```
5. add annotation @RolesAllowed(<role>) to your REST methods

**Reply by mariuszbozek** [reply]
try this one https://www.kantega.no/blogg/role-based-access-control-with-quarkus-and-microsoft-azure-active-directory

**Reply by sberyozkin** [reply]
It is possible that the token which is sent in this case is in JWT format this is why it works, might depend in Azure version. Quarkus will fetch the public keys from `https://login.microsoftonline.com/<tenant>/v2.0` so if it is an access token in a binary format then they won't work, so in such cases  `https://sts.windows.net/<tenant-id>` may help

**Reply by dkoci** [reply]
Do you have any idea how to invoke azure secured api using swagger openapi?

**Reply by mariuszbozek** [reply]
> Do you have any idea how to invoke azure secured api using swagger openapi?

Please try the annotation @SecuritySchemas and then the annotation @SecurityRequirement

https://swagger.io/docs/specification/authentication/oauth2/

**Reply by dkoci** [reply]
I dont know what happened, but I cant access my backend api anymore from angular frontend app. I tried both oidc-auth-url (with access and id token) but it is not working. Get me 401 response code.

**Reply by sberyozkin** [reply]
May be some stale state in the browser or a limit to a number of tokens Azure can issue. Hard to tell, try

https://quarkus.io/guides/security-openid-connect#how-to-check-the-errors-in-the-logs

with 2.9.0.Final/2.8.2.Final

**Reply by dkoci** [reply]
I found what was wrong. In my previous app I tested role based access. Had `quarkus.oidc.roles.role-claim-path=roles` in application.properties (default is groups, but azure uses roles so we have to tell quarkus to use roles instead of groups). I had to remove this line from application.properties because there were no roles field in token.

**Reply by dkoci** [reply]
Have another problem :) Health probe gives me an error: Error processing Health Checks: io.quarkus.security.UnauthorizedException
There is also some strange health check added: <package_name>.health.ReadinesProbe_ClientProxy
Is there any way to disable authentication for health checks?

**Reply by sberyozkin** [reply]
Np, it would make sense to keep this thread to Azure specific queries, but indeed, you can permit all `/q/*` queries:
```
quarkus.http.auth.permission.public.paths=/q/*
quarkus.http.auth.permission.public.policy=permit
```

See https://quarkus.io/guides/security-authorization

**Reply by dkoci** [reply]
Not working. The problem is my health check calls some method which has restricted access.

**Reply by sberyozkin** [reply]
In that case it does not matter whether it is a health check or not, it is about either requiring a protected access to a given method or not.
Unfortunately it is not possible to have the same method semi-secured - where one can check for permissions for all requests to this method but the health checks
